[FLINK-12706] Introduce ShuffleService interface and its configuration

azagrebin opened the pull request:
## What is the purpose of the change

Introduce `ShuffleService` interface with two factory methods for `ShuffleMaster` and `ShuffleEnvironment` and its configuration. Implement and configure by default `NettyShuffleService`.

## Brief change log

  - Introduce `ShuffleService` interface with two factory methods for `ShuffleMaster` and `ShuffleEnvironment`
  - Introduce configuration core option with the full name of the implementation class and a `ShuffleServiceLoader` utility class to load the configured `ShuffleService` implementation from the class path 
  - Implement and configure by default `NettyShuffleService`.

## Verifying this change

Refactoring, existing unit tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (np)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (docs / JavaDocs)

azagrebin commented:
The PR is based on #8608 at the moment, only last two commits are to review.

azagrebin commented:
@flinkbot attention @tillrohrmann @zhijiangW

zhijiangW commented on flink-core/src/main/java/org/apache/flink/configuration/CoreOptions.java:
```diff
@@ -306,4 +306,17 @@
+		.key("shuffle.service.class.name")
+		.defaultValue("org.apache.flink.runtime.io.network.NettyShuffleService")
+		.withDescription("The full name of shuffle service implementation class. " +
+			"The default implentation uses netty network communication, local memory and file system on task executor");
```
implentation -> implementation

zhijiangW commented on flink-core/src/main/java/org/apache/flink/configuration/CoreOptions.java:
```diff
@@ -306,4 +306,17 @@
+	/**
+	 * The shuffle service implementation class full name.
+	 */
+	public static final ConfigOption<String> SHUFFLE_SERVICE = ConfigOptions
```
why not put this option into runtime module? Are there any considerations

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleService.java:
```diff
@@ -0,0 +1,124 @@
+			TaskEventPublisher taskEventPublisher,
+			MetricGroup metricGroup,
+			IOManager ioManager) {
+		checkNotNull(ioManager);
```
better to obey the parameters sequence during `checkNotNull`.
also check for the left `taskExecutorLocation` and `metricGroup`?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleService.java:
```diff
@@ -0,0 +1,124 @@
+	@VisibleForTesting
+	static NettyShuffleEnvironment createNettyShuffleEnvironment(
+			NettyShuffleEnvironmentConfiguration config,
+			ResourceID taskExecutorLocation,
```
taskExecutorLocation -> resourceID, because we have the `TaskManagerLocation` class which might be confused

tillrohrmann replied:
Could also be called `taskExecutorResourceId`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleService.java:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	private static void registerNetworkMetrics(MetricGroup metricGroup, NetworkBufferPool networkBufferPool) {
+		checkNotNull(metricGroup);
```
might remove this if this check could be done before in `createNettyShuffleEnvironment`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java:
```diff
@@ -163,4 +167,48 @@
+	/**
+	 * Task executor local context of shuffle service used to create {@link ShuffleEnvironment}.
+	 */
+	class ShuffleLocalContext {
```
This class seems not be used atm. What is the motivation for bringing this?

azagrebin replied:
true, some rebase leftover, it is now `ShuffleEnvironmentContext`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleService.java:
```diff
@@ -18,18 +18,28 @@
+	 * Factory method to create a specific {@link ShuffleMaster} implementation in job master.
+	 *
+	 * @param configuration Flink configuration
+	 * @return shuffle manager implementation
```
shuffle manager -> shuffle master

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleServiceLoader.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Utility to load the pluggable {@link ShuffleService} implementations.
+ */
+public class ShuffleServiceLoader {
```
It might need a unit test for covering this class

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java:
```diff
@@ -170,6 +171,8 @@
 
 	private final Map<ResourceID, Tuple2<TaskManagerLocation, TaskExecutorGateway>> registeredTaskManagers;
 
+	private final ShuffleMaster<?> shuffleMaster;
```
Why do we store the `ShuffleMaster` here? Wouldn't it be enough to pass it on to the `createScheduler` method?

azagrebin replied:
`createScheduler` is currently called not only from the constructor

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleService.java:
```diff
@@ -0,0 +1,51 @@
+ * @param <P> type of provided result partition writers
+ * @param <G> type of provided input gates
+ */
+public interface ShuffleService<SD extends ShuffleDescriptor, P extends ResultPartitionWriter, G extends InputGate> {
```
Maybe `ShuffleServiceFactory` would better describe what it does.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleService.java:
```diff
@@ -0,0 +1,51 @@
+public interface ShuffleService<SD extends ShuffleDescriptor, P extends ResultPartitionWriter, G extends InputGate> {
+
+	/**
+	 * Factory method to create a specific {@link ShuffleMaster} implementation in job master.
```
remove `in job master`

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleService.java:
```diff
@@ -0,0 +1,51 @@
+	ShuffleMaster<SD> createShuffleMaster(Configuration configuration);
+
+	/**
+	 * Factory method to create a specific local {@link ShuffleEnvironment} implementation in task executor.
```
remove `in task executor`

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleServiceLoader.java:
```diff
@@ -0,0 +1,58 @@
+		return shuffleService.orElseThrow(() -> new FlinkException(
+			String.format(
+				"Could not instantiate the ShuffleService '%s'. " +
+					"Please make sure that this class is on your class path.",
```
Indentation seems wrong.

azagrebin replied:
This is still the same argument, the string is broken by concatenation, I suggest we leave it this way to distinguish it from the next argument.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleServiceLoader.java:
```diff
@@ -0,0 +1,58 @@
+/**
+ * Utility to load the pluggable {@link ShuffleService} implementations.
+ */
+public class ShuffleServiceLoader {
```
Could simply be an enum since it only contains static methods.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleServiceLoader.java:
```diff
@@ -0,0 +1,58 @@
+				ShuffleService.class,
+				classLoader));
+		} catch (ClassNotFoundException e) {
+			return Optional.empty();
```
Why are we swallowing this exception here? It looks as if we recreate it at the call site of `lookupEmbeddedShuffleServices` if the result is `empty`. I think it would be better to simply return `ShuffleService<?, ?, ?>` and to allow this method to throw a `ClassNotFoundException`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleServiceLoader.java:
```diff
@@ -0,0 +1,58 @@
+				shuffleServiceClassName)));
+	}
+
+	private static Optional<ShuffleService<?, ?, ?>> lookupEmbeddedShuffleServices(String shuffleServiceClassName) {
```
This method has some code duplication with `HighAvailabilityServiceUtils#createCustomHAServices`. Maybe we could factor a common method out which calls `InstantiationUtil.instantiate` and creates a `FlinkException` in case of a `ClassNotFoundException`.

azagrebin replied:
I will move wrapping of `ClassNotFoundException` by  `FlinkException` into `InstantiationUtil.instantiate`.

azagrebin commented:
Thanks for the review @zhijiangW @tillrohrmann, I have addressed comments

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java:
```diff
@@ -50,10 +50,11 @@
  */
 public class SingleInputGateFactory {
+	@SuppressWarnings("LoggerInitializedWithForeignClass")
 	private static final Logger LOG = LoggerFactory.getLogger(SingleInputGate.class);
```
why not use `SingleInputGateFactory.class` here?

azagrebin replied:
My idea was to preserve the structure of the previous logging messages because the factory was a result of refactoring. I can change this if it is not important.

zhijiangW replied:
got it, thanks for the explanation!

zhijiangW commented on docs/ops/config.md:
```diff
@@ -132,6 +132,10 @@ The configuration keys in this section are independent of the used resource mana
 
 {% include generated/resource_manager_configuration.html %}
 
+### Shuffle service
```
Upper case for `Service` for consistent with others.

zhijiangW commented on docs/ops/config.zh.md:
```diff
@@ -132,6 +132,10 @@ The configuration keys in this section are independent of the used resource mana
 
 {% include generated/resource_manager_configuration.html %}
 
+### Shuffle service
```
ditto: `Service`

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/shuffle/ShuffleServiceLoaderTest.java:
```diff
@@ -0,0 +1,80 @@
+import org.apache.flink.runtime.io.network.api.writer.ResultPartitionWriter;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.util.FlinkException;
+import org.junit.Test;
```
should add an empty line before this import.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/shuffle/ShuffleServiceLoaderTest.java:
```diff
@@ -0,0 +1,80 @@
+		configuration.setString(SHUFFLE_SERVICE_FACTORY_CLASS, "org.apache.flink.runtime.shuffle.ShuffleServiceLoaderTest$CustomShuffleServiceFactory");
+		ShuffleServiceFactory<?, ?, ?> shuffleServiceFactory = ShuffleServiceLoader.loadShuffleServiceFactory(configuration);
+		assertThat(
+			"Loaded shuffle service factory is not the default netty implementation",
```
should adjust the message

zhijiangW commented:
@azagrebin thanks for the updates and I only left several minor comments.

BTW, the commit of `[hotfix][network] Rename taskExecutorLocation` should be tagged as `fixup` instead of `hotfix`. `fixup` would be squashed when merging, but `hotfix` would be merged in a separate commit.

azagrebin commented:
@zhijiangW 
thanks for the  review, I have addressed comments.
I actually planned `[hotfix][network] Rename taskExecutorLocation` as a separate commit because it does somewhat unrelated cleanup.

zhijiangW commented:
Sorry I thought the renaming issue was caused by this PR in previous commit before. 
Yes, if this renaming already existed before this PR, it should be a `hotfix` here. :)
Thanks for updating, LGTM!

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironment.java:
```diff
@@ -358,6 +286,7 @@ public void close() {
 			LOG.info("Shutting down the network environment and its components.");
 
 			// terminate all network connections
+			//noinspection OverlyBroadCatchBlock
```
Let's revert this.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/shuffle/ShuffleServiceLoaderTest.java:
```diff
@@ -0,0 +1,81 @@
+/**
+ * Test suite for {@link ShuffleServiceLoader} utility.
+ */
+public class ShuffleServiceLoaderTest {
```
`extends TestLogger` is missing
